ffi/src/expressions/kernel_expr_visitor.rs

```diff
@@ -0,0 +1,354 @@
+    /// The operands will be in a _two_ item list identified by `child_list_id`
+    pub visit_in: VisitBinaryOpFn,
+    /// Visits the `NotIn` binary operator belonging to the list identified by `sibling_list_id`.
+    /// The operands will be in a _two_ item list identified by `child_list_id`
+    pub visit_not_in: VisitBinaryOpFn,
+    /// Visits the `Add` binary operator belonging to the list identified by `sibling_list_id`.
+    /// The operands will be in a _two_ item list identified by `child_list_id`
+    pub visit_add: VisitBinaryOpFn,
+    /// Visits the `Minus` binary operator belonging to the list identified by `sibling_list_id`.
+    /// The operands will be in a _two_ item list identified by `child_list_id`
+    pub visit_minus: VisitBinaryOpFn,
+    /// Visits the `Multiply` binary operator belonging to the list identified by `sibling_list_id`.
+    /// The operands will be in a _two_ item list identified by `child_list_id`
+    pub visit_multiply: VisitBinaryOpFn,
+    /// Visits the `Divide` binary operator belonging to the list identified by `sibling_list_id`.
+    /// The operands will be in a _two_ item list identified by `child_list_id`
+    pub visit_divide: VisitBinaryOpFn,
+    /// Visits the `column` belonging to the list identified by `sibling_list_id`.
+    pub visit_column:
+        extern "C" fn(data: *mut c_void, sibling_list_id: usize, name: KernelStringSlice),
+    /// Visits a `StructExpression` belonging to the list identified by `sibling_list_id`.
+    /// The sub-expressions of the `StructExpression` are in a list identified by `child_list_id`
+    pub visit_struct_expr:
+        extern "C" fn(data: *mut c_void, sibling_list_id: usize, child_list_id: usize),
+}
+
+/// Visit the expression of the passed [`SharedExpression`] Handle using the provided `visitor`.
+/// See the documentation of [`EngineExpressionVisitor`] for a description of how this visitor
+/// works.
+///
+/// This method returns the id that the engine generated for the top level expression
+///
+/// # Safety
+///
+/// The caller must pass a valid SharedExpression Handle and expression visitor
+#[no_mangle]
+pub unsafe extern "C" fn visit_expression(
+    expression: &Handle<SharedExpression>,
+    visitor: &mut EngineExpressionVisitor,
+) -> usize {
+    macro_rules! call {
+        ( $visitor:ident, $visitor_fn:ident $(, $extra_args:expr) *) => {
+            ($visitor.$visitor_fn)($visitor.data $(, $extra_args) *)
+        };
+    }
+    fn visit_array(
+        visitor: &mut EngineExpressionVisitor,
+        array: &ArrayData,
+        sibling_list_id: usize,
+    ) {
+        #[allow(deprecated)]
+        let elements = array.array_elements();
+        let child_list_id = call!(visitor, make_field_list, elements.len());
+        for scalar in elements {
+            visit_scalar(visitor, scalar, child_list_id);
+        }
+        call!(visitor, visit_array_literal, sibling_list_id, child_list_id);
+    }
+    fn visit_struct_literal(
```

Comment:
hrmm, this is a tricky one. I don't _love_ converting the field names into string scalars and maintaining two lists.

We already have a way of doing iterators, so what if we visit all the values, and then `visit_struct_literal` could take an iterator over `KernelStringSlice` that will return field names, and the list id where the engine can find the value for each field.

I don't _love_ that solution either, but it feels more clean than the engine having to unwrap all these string scalars into just strings. The downside is that you're pairing an iterator to a list, so the engine has to deal with potentially two different ways of iterating.

I realize this might all change since we're potentially doing away with/seriously changing `StructData`, so if you think it's not worth tackling until then I'm okay with this in the short term.

also aside: can we rename this internal function so it doesn't have the same name as the visitor one, it's confusing :)

Reply:
This was intended as a quick placeholder from the beginning because `StructField` would've been better to pass. On top of that we're planning a major rework of scalars/structs/arrays, so I think this is okay as a first draft.

Reply:
Renamed kernel visitor functions to `visit_expression_*` 👍